internal/users/idgenerator.go

```diff
@@ -0,0 +1,265 @@
+package users
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"slices"
+	"sort"
+
+	"github.com/ubuntu/authd/internal/users/localentries"
+	"github.com/ubuntu/authd/log"
+)
+
+// IDGeneratorIface is the interface that must be implemented by the ID generator.
+type IDGeneratorIface interface {
+	GenerateUID(ctx context.Context, owner IDOwner) (uid uint32, cleanup func(), err error)
+	GenerateGID(ctx context.Context, owner IDOwner) (gid uint32, cleanup func(), err error)
+}
+
+// IDOwner is the interface that must be implemented by the IDs owner to provide
+// the currently used UIDs and GIDs.
+type IDOwner interface {
+	UsedUIDs() ([]uint32, error)
+	UsedGIDs() ([]uint32, error)
+}
+
+// IDGenerator is an ID generator that generates UIDs and GIDs in a specific range.
+type IDGenerator struct {
+	UIDMin uint32
+	UIDMax uint32
+	GIDMin uint32
+	GIDMax uint32
+
+	// IDs generated but not saved to the database yet.
+	// This is used to avoid generating the same ID multiple times.
+	// We don't differentiate between UIDs and GIDs here, because:
+	// * When picking a UID, we avoid IDs which are already used as GIDs,
+	//   because the UID is also used as the GID of the user private group.
+	// * When picking a GID, we avoid IDs which are already used as UIDs,
+	//   because those are also GIDs of the user private groups.
```

Comment:
Let's make this more precise. We currently don't avoid picking ID candidates which other NSS sources might already use (we check those in `isAvailableID` though) and the part "UID is also used as the GID of the user private group" is also only guaranteed for authd users.
```suggestion
	// * When picking a UID, we avoid IDs which we already used as GIDs,
	//   because the UID is also used as the GID of the user private group.
	// * When picking a GID, we avoid IDs which we already used as UIDs,
	//   because those are also GIDs of the user private groups.
```